app/shared/model/base.model.ts

```diff
@@ -0,0 +1,210 @@
+        this.init(this.parent, doc, false);
+      }
+
+      // save related docs in parallel
+      let relatedSavers: Observable<boolean>[] = this.getRelated().filter((model) => {
+        return model.changed();
+      }).map((model) => {
+        model.parent = this.doc; // update reference to haldoc
+        return model.save();
+      });
+      return Observable.fromArray(relatedSavers).concatAll().toArray().map(() => {
+        this.isNew = false;
+        this.isSaving = false;
+        return true;
+      });
+    });
+  }
+
+  changed(field?: string | string[]): boolean {
+    let fields = (field instanceof Array) ? field : [field];
+    if (!field || field.length < 1) {
+      fields = this.SETABLE.concat(this.RELATIONS);
+    }
+    for (let f of fields) {
+      if (this.RELATIONS.indexOf(f) > -1) {
+        for (let model of this.getRelated(f)) {
+          if (model.changed()) { return true; }
+        }
+      } else if (this.changedFields[f]) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  invalid(field?: string | string[]): string {
+    let fields = (field instanceof Array) ? field : [field];
+    if (!field || field.length < 1) {
+      fields = this.SETABLE.concat(this.RELATIONS);
+    }
+    let invalids: string[] = [];
+    for (let f of fields) {
+      if (this.RELATIONS.indexOf(f) > -1) {
+        for (let model of this.getRelated(f)) {
+          if (model.invalid()) { invalids.push(model.invalid()); }
+        }
+      } else if (this.invalidFields[f]) {
+        invalids.push(this.invalidFields[f]);
+      }
+    }
+    return invalids.length ? invalids.join(', ') : null;
+  }
+
+  store() {
+    if (window && window.localStorage && this.key()) {
+      let changed = {};
+      for (let f of Object.keys(this.changedFields)) {
+        changed[f] = this[f];
+      }
+      window.localStorage.setItem(this.key(), JSON.stringify(changed));
+    }
+  }
+
+  restore() {
+    if (window && window.localStorage && this.key()) {
+      let json = window.localStorage.getItem(this.key());
+      if (json) {
+        let data = JSON.parse(json);
+        for (let key of Object.keys(data)) {
+          if (this.SETABLE.indexOf(key) > -1) {
+            this[key] = data[key];
+          }
+        }
+      }
+    }
+  }
+
+  unstore() {
+    if (window && window.localStorage && this.key()) {
+      window.localStorage.removeItem(this.key());
+    }
+  }
+
+  revalidate() {
+    for (let f of this.SETABLE) {
+      this.invalidFields[f] = this.invalidate(f, this[f]);
+      this.changedFields[f] = this.changidate(f, this[f]);
+    }
+  }
+
+  invalidate(field: string, value: any): string {
+    let validators = this.VALIDATORS[field] || [];
+    for (let validator of validators) {
+      let invalidMsg = validator(field, value);
+      if (invalidMsg) {
+        return invalidMsg;
+      }
+    }
+    return null;
+  }
+
+  changidate(field: string, value: any): boolean {
```

Comment:
Checking if a field has changed from the original (remotely-stored) value.

Naming is hard, okay?  But yes, something like `changecheck` would probably be better than my sarcastic naming.